march_hardware_interface/src/march_hardware_interface.cpp

```diff
@@ -460,20 +470,59 @@ bool MarchHardwareInterface::iMotionCubeStateCheck(size_t joint_index)
-                       joint.getName().c_str(), soft_limits_[joint_index].min_position,
-                       soft_limits_[joint_index].max_position, joint_position_[joint_index]);
-
-    if (joint.canActuate())
+    if (joint_position_[joint_index] < soft_limits_error_[joint_index].min_position ||
+        joint_position_[joint_index] > soft_limits_error_[joint_index].max_position)
     {
-      std::ostringstream error_stream;
-      error_stream << "Joint " << joint.getName() << " is out of its soft limits ("
-                   << soft_limits_[joint_index].min_position << ", " << soft_limits_[joint_index].max_position
-                   << "). Actual position: " << joint_position_[joint_index];
-      throw std::runtime_error(error_stream.str());
+      ROS_ERROR_THROTTLE(1, "Joint %s is outside of its error soft limits (%f, %f). Actual position: %f",
+                         joint.getName().c_str(), soft_limits_error_[joint_index].min_position,
+                         soft_limits_error_[joint_index].max_position, joint_position_[joint_index]);
+
+      if (joint.canActuate())
+      {
+        std::ostringstream error_stream;
+        error_stream << "Joint " << joint.getName() << " is out of its soft limits ("
+                     << soft_limits_[joint_index].min_position << ", " << soft_limits_[joint_index].max_position
+                     << "). Actual position: " << joint_position_[joint_index];
+        throw std::runtime_error(error_stream.str());
+      }
     }
+
+    ROS_WARN_THROTTLE(1, "Joint %s is outside of its soft limits (%f, %f). Actual position: %f",
+                      joint.getName().c_str(), soft_limits_[joint_index].min_position,
+                      soft_limits_[joint_index].max_position, joint_position_[joint_index]);
   }
 }
+
+void MarchHardwareInterface::getSoftJointLimitsError(const std::string& name,
+                                                     const urdf::JointConstSharedPtr& urdf_joint,
+                                                     joint_limits_interface::SoftJointLimits& error_soft_limits)
+{
+  std::ostringstream param_name;
+  std::ostringstream error_stream;
+
+  param_name << "/march/controller/trajectory/constraints/" << name << "/percent_soft_limit_error";
+
+  if (!ros::param::has(param_name.str()))
+  {
+    error_stream << "Margin soft limits error of joint: " << name << " could not be found";
+    throw std::runtime_error(error_stream.str());
+  }
+
+  float margin;
+  ros::param::param<float>(param_name.str(), margin, 0.0);
+
+  if (!urdf_joint || !urdf_joint->safety || !urdf_joint->limits || margin == 0.0)
+  {
+    error_stream << "Could not construct the soft limits for joint: " << name;
+    throw std::runtime_error(error_stream.str());
+  }
+
+  error_soft_limits.min_position =
+      urdf_joint->limits->lower + ((urdf_joint->limits->lower - urdf_joint->safety->soft_lower_limit) * margin);
```

Comment:
I think this should be the other way around.
```suggestion
      urdf_joint->limits->lower + ((urdf_joint->safety->soft_lower_limit - urdf_joint->limits->lower) * margin);
```

Reply:
copy error :see_no_evil: